core/src/main/java/ai/timefold/solver/core/impl/constructionheuristic/placer/QueuedMultiplePlacer.java

```diff
@@ -0,0 +1,213 @@
+        private Move<Solution_> cachedMove = null;
+
+        private MultipleQueuedPlacingIterator(List<EntityPlacer<Solution_>> queuedPlacerList) {
+            this.queuedPlacerList = new ArrayList<>(queuedPlacerList);
+            reset();
+        }
+
+        /**
+         * The method uses a strategy similar to {@link CartesianProductMoveSelector},
+         * but it uses placer iterators instead.
+         */
+        private Move<Solution_> nextMove() {
+            if (cachedMove == null) {
+                var childSize = moveIteratorList.size();
+                int index;
+                Move<Solution_>[] move = new Move[childSize];
+                if (previousMove == null) {
+                    index = -1;
+                } else {
+                    index = consumeNextMove(move, previousMove);
+                    if (index == -1) {
+                        // No more moves
+                        return null;
+                    }
+                }
+                var recreated = recreateNextIterators(index, move);
+                if (!recreated) {
+                    // We stop if one of the placement iterators has no next placement
+                    return null;
+                }
+                previousMove = move;
+                cachedMove = CompositeMove.buildMove(move);
+            }
+            return cachedMove;
+        }
+
+        /**
+         * Go through the registered iterators and check for any available moves.
+         *
+         * @param move the move array to be loaded
+         * @param previousMove the previous moves
+         * @return the last index of the iterator that still has available moves; otherwise, the function returns -1
+         *         when all iterators have no more moves available.
+         */
+        private int consumeNextMove(Move<?>[] move, Move<Solution_>[] previousMove) {
+            var index = move.length - 1;
+            // Look for the first iterator that still has available moves to generate
+            while (index >= 0) {
+                var moveIterator = moveIteratorList.get(index);
+                if (moveIterator.hasNext()) {
+                    break;
+                }
+                index--;
+            }
+            if (index < 0) {
+                return -1;
+            }
+            // Copy the previous move until the next one generated
+            System.arraycopy(previousMove, 0, move, 0, index);
+            // Generate and set the new move
+            move[index] = moveIteratorList.get(index).next();
+            return index;
+        }
+
+        /**
+         * Recreate all move iterators starting from #lastValidIteratorIndex.
+         *
+         * @param lastValidIteratorIndex the index of the last iterator that generated a valid move
+         * @param move the move array to be loaded
+         */
+        private boolean recreateNextIterators(int lastValidIteratorIndex, Move<?>[] move) {
```

Comment:
Let's avoid side effects.
The method can return an array on success, and null on failure.
That way, it can create the array internally, keeping logic more focused and isolated.